### PR TITLE
Forward `shuffle` to trainer server and measure GPU

### PR DESCRIPTION
**Pull request by XianzheMa**
1. Add a unit test to how we pack the `StartTrainingRequest`, thereby catching the unspecified `shuffle`.
2. Enable accurate GPU operations measurement.

### CODE REVIEW

File: modyn/common/grpc/grpc_helpers.py [hunk 1]
```diff
@@ -244,23 +232,46 @@ def start_training(
...
-            "num_prefetched_partitions": num_prefetched_partitions,
-            "parallel_prefetch_requests": parallel_prefetch_requests,
-            "seed": seed,
+            "epochs_per_trigger": training_config.epochs_per_trigger,
+            "num_prefetched_partitions": training_config.num_prefetched_partitions,
+            "parallel_prefetch_requests": training_config.parallel_prefetch_requests,
+            "seed": training_config.seed,
             "tokenizer": PythonString(value=tokenizer) if tokenizer is not None else None,
             "num_samples_to_pass": num_samples_to_pass,
+            "shuffle": training_config.shuffle,
```
Comment (XianzheMa) [inline modyn/common/grpc/grpc_helpers.py:246]: @MaxiBoether if you urgently need to run experiments, this is the only line missing to currectly set `shuffle` :D

Comment (MaxiBoether) [reply modyn/common/grpc/grpc_helpers.py:246]: Nah, I need to wait until we fix the most recent model stuff anyways. I will probably run experiments again over night :D

File: modyn/common/grpc/grpc_helpers.py [hunk 2]
```diff
@@ -207,15 +203,7 @@ def start_training(
             lr_scheduler_configs = training_config.lr_scheduler.model_dump(by_alias=True)
 
         criterion_config = json.dumps(training_config.optimization_criterion.config)
-
-        epochs_per_trigger = training_config.epochs_per_trigger
-        num_prefetched_partitions = training_config.num_prefetched_partitions
-        parallel_prefetch_requests = training_config.parallel_prefetch_requests
-
```
Comment (XianzheMa) [inline modyn/common/grpc/grpc_helpers.py:214]: to reduce the number of local variables

File: modyn/trainer_server/internal/trainer/maybe_measure_gpu_ops.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
+from types import TracebackType
+from typing import Optional, Type
+
+import torch
+from modyn.common.benchmark import Stopwatch
+
+
+class MaybeMeasureGPUOps:
+    def __init__(  # type: ignore[no-untyped-def]
+        self, measure: bool, measurement_name: str, device: str, stop_watch: Stopwatch, **kwargs
```
Comment (MaxiBoether) [inline modyn/trainer_server/internal/trainer/maybe_measure_gpu_ops.py:10]: nit: can we rename measure to enabled? that is the common name for things like this I believe, at least I have seen this pattern often

File: modyn/trainer_server/internal/trainer/maybe_measure_gpu_ops.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
+from types import TracebackType
+from typing import Optional, Type
+
+import torch
+from modyn.common.benchmark import Stopwatch
+
+
+class MaybeMeasureGPUOps:
```
Comment (MaxiBoether) [inline modyn/trainer_server/internal/trainer/maybe_measure_gpu_ops.py:8]: nit: similar to my comment below, let's name this `GPUMeasurement` and let it have an enabled flag, to make the name less convoluted

File: modyn/trainer_server/internal/trainer/maybe_measure_gpu_ops.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+        self._device = device
+        self._measure = measure
+        self._measurement_name = measurement_name
+        self._kwargs = kwargs
+        self._stop_watch = stop_watch
+
+    def __enter__(self) -> None:
+        if self._measure:
+            torch.cuda.synchronize(device=self._device)
+            self._stop_watch.start(name=self._measurement_name, **self._kwargs)
```
Comment (MaxiBoether) [inline modyn/trainer_server/internal/trainer/maybe_measure_gpu_ops.py:21]: Can we always measure the stopwatch, and have only cuda synchronize wrapped in the if statement? I understand the numbers might not make full sense but in the past I still sometimes took a look to get an indication whether _something_ is happening. Thanks!

**Review by MaxiBoether — CHANGES_REQUESTED** [review]
Thanks Xianzhe! I had a nitpicky naming comment, and I would like to always have the stopwatch running, even if we don't properly synchronize cuda.

File: modyn/trainer_server/internal/trainer/maybe_measure_gpu_ops.py (hunk 5 above)
Comment (XianzheMa) [reply modyn/trainer_server/internal/trainer/maybe_measure_gpu_ops.py]: done

File: modyn/trainer_server/internal/trainer/maybe_measure_gpu_ops.py (hunk 4 above)
Comment (XianzheMa) [reply modyn/trainer_server/internal/trainer/maybe_measure_gpu_ops.py]: done

File: modyn/trainer_server/internal/trainer/maybe_measure_gpu_ops.py (hunk 3 above)
Comment (XianzheMa) [reply modyn/trainer_server/internal/trainer/maybe_measure_gpu_ops.py]: done

**Review by MaxiBoether — APPROVED** [review]
Looks good!
